File: project/hyeok-onboarding/src/main/java/com/INNER_CIRCLE_ICD4/innerCircle/service/SurveyService.java
package com.INNER_CIRCLE_ICD4.innerCircle.service;

import com.INNER_CIRCLE_ICD4.innerCircle.domain.Choice;
import com.INNER_CIRCLE_ICD4.innerCircle.domain.Question;
import com.INNER_CIRCLE_ICD4.innerCircle.domain.Survey;
import com.INNER_CIRCLE_ICD4.innerCircle.dto.*;
import com.INNER_CIRCLE_ICD4.innerCircle.repository.SurveyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class SurveyService {

    private final SurveyRepository surveyRepository;

    @Transactional
    public SurveyResponse createSurvey(SurveyRequest req) {
        Survey survey = new Survey(req.title(), req.description());

        for (QuestionRequest qr : req.questions()) {
            Question q = new Question(
                    qr.title(),
                    qr.description(),
                    qr.type(),
                    qr.required()
            );
            survey.addQuestion(q);

            if (qr.choices() != null) {
                for (int i = 0; i < qr.choices().size(); i++) {
                    q.addChoice(new Choice(qr.choices().get(i), i));
                }
            }
        }

        Survey saved = surveyRepository.save(survey);
        return toDto(saved);
    }

    @Transactional(readOnly = true)
    public SurveyResponse findById(UUID id) {
        Survey survey = surveyRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "해당 설문을 찾을 수 없습니다: " + id));
        return toDto(survey);
    }

    @Transactional
    public SurveyResponse update(UUID id, SurveyUpdateRequest req) {
        Survey survey = surveyRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "해당 설문을 찾을 수 없습니다: " + id));

        // 제목과 설명 업데이트
        survey.update(req.title(), req.description());

        // 질문 교체
        List<Question> newQuestions = req.questions().stream().map(qur -> {
            Question q = new Question(
                    qur.title(),
                    qur.description(),
                    qur.type(),
                    qur.required()
            );
            if (qur.choices() != null) {
                for (int i = 0; i < qur.choices().size(); i++) {
                    q.addChoice(new Choice(qur.choices().get(i), i));
                }
            }
            return q;
        }).collect(Collectors.toList());

        survey.replaceQuestions(newQuestions);
        return toDto(survey);
    }

    private SurveyResponse toDto(Survey s) {
        List<QuestionResponse> qs = s.getQuestions().stream()
                .map(q -> new QuestionResponse(
                        q.getId(),
                        q.getTitle(),
                        q.getDescription(),
                        q.getType(),
                        q.isRequired(),
                        q.getChoices().stream()
                                .map(c -> new ChoiceResponse(
                                        c.getId(),
                                        c.getText(),
                                        c.getChoiceIndex()
                                ))
                                .collect(Collectors.toList())
                ))
                .collect(Collectors.toList());

        return new SurveyResponse(
                s.getId(),
                s.getTitle(),
                s.getDescription(),
                s.getVersion(),
                qs
        );
    }
}

File: project/jinho-onboarding/src/main/kotlin/fc/innercircle/jinhoonboarding/survey/service/SurveyService.kt
package fc.innercircle.jinhoonboarding.survey.service

import fc.innercircle.jinhoonboarding.survey.dto.request.SurveyRequest
import fc.innercircle.jinhoonboarding.survey.dto.response.SurveyResponse
import fc.innercircle.jinhoonboarding.survey.repository.SurveyRepository
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class SurveyService(
    private val surveyRepository: SurveyRepository
) {

    fun getSurvey(surveyId: Long): SurveyResponse {
        return SurveyResponse.fromDomain(surveyRepository.findById(surveyId).orElseThrow{ RuntimeException("설문 ID가  $surveyId 인 설문은 존재하지 않습니다.") })
    }

    fun createSurvey(request: SurveyRequest): SurveyResponse {
        val surveyToCreate = request.toDomain()
        return SurveyResponse.fromDomain(surveyRepository.save(surveyToCreate))
    }
    @Transactional
    fun updateSurvey(request: SurveyRequest, surveyId: Long): SurveyResponse {
        // 수정 할 대상 설문을 가져온다
        val surveyToUpdate = surveyRepository.findById(surveyId).orElseThrow{ RuntimeException("수정할 설문이 없습니다: $surveyId") }

        val tobeSurvey = request.toDomain()
        val questions = tobeSurvey.questions

        surveyToUpdate.updateTitle(tobeSurvey.title)
        surveyToUpdate.updateDescription(tobeSurvey.description)
        surveyToUpdate.updateQuestions(questions)

        return SurveyResponse.fromDomain(surveyRepository.save(surveyToUpdate))
    }

    fun deleteById(surveyId: Long) {
        surveyRepository.findById(surveyId).orElseThrow{ RuntimeException("삭제할 설문이 없습니다: $surveyId") }
        surveyRepository.deleteById(surveyId)
    }

}